internal/resources/gsi.go

```diff
@@ -0,0 +1,509 @@
+	_ resource.ResourceWithValidateConfig = (*GSI)(nil)
+)
+
+// GSI is the GSI resource implementation.
+type GSI struct {
+	*providerschema.Data
+}
+
+func NewGSI() resource.Resource {
+	return &GSI{}
+}
+
+// Metadata returns the query index resource type name.
+func (g *GSI) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_query_indexes"
+}
+
+// Schema defines the schema for the query index resource.
+func (g *GSI) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = GsiSchema()
+}
+
+// Create will create/drop/build/alter a primary or secondary index.
+func (g *GSI) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var ddl string
+
+	// create build index statement.
+	if !plan.BuildIndexes.IsNull() {
+		ddl = fmt.Sprintf(
+			"BUILD INDEX ON `%s`.`%s`.`%s`(%s)",
+			plan.BucketName.ValueString(),
+			plan.ScopeName.ValueString(),
+			plan.CollectionName.ValueString(),
+			listStringValues(plan.BuildIndexes),
+		)
+	} else {
+		// create primary index statement.
+		if plan.IsPrimary.ValueBool() {
+			var indexName string
+			if !plan.IndexName.IsNull() {
+				indexName = plan.IndexName.ValueString()
+			} else {
+				indexName = "#primary"
+			}
+			ddl = fmt.Sprintf(
+				"CREATE PRIMARY INDEX `%s` ON `%s`.`%s`.`%s`  WITH { \"defer_build\": %t,  \"num_replica\": %d }",
+				indexName,
+				plan.BucketName.ValueString(),
+				plan.ScopeName.ValueString(),
+				plan.CollectionName.ValueString(),
+				plan.With.DeferBuild.ValueBool(),
+				plan.With.NumReplica.ValueInt64(),
+			)
+		} else {
+			// create secondary index statement.
+			ddl = fmt.Sprintf(
+				"CREATE INDEX `%s` ON `%s`.`%s`.`%s`(%s) ",
+				plan.IndexName.ValueString(),
+				plan.BucketName.ValueString(),
+				plan.ScopeName.ValueString(),
+				plan.CollectionName.ValueString(),
+				listStringValues(plan.IndexKeys),
+			)
+
+			if !plan.PartitionBy.IsNull() {
+				ddl += fmt.Sprintf(" PARTITION BY HASH(%s) ", listStringValues(plan.PartitionBy))
+			}
+
+			if !plan.Where.IsNull() {
+				ddl += fmt.Sprintf(" WHERE %s ", plan.Where.ValueString())
+			}
+
+			if !plan.PartitionBy.IsNull() {
+				withClause := fmt.Sprintf(
+					" WITH { \"defer_build\": %t,  \"num_replica\": %d,  \"num_partition\": %d } ",
+					plan.With.DeferBuild.ValueBool(),
+					plan.With.NumReplica.ValueInt64(),
+					plan.With.NumPartition.ValueInt64(),
+				)
+
+				ddl += withClause
+			} else {
+				// should not set num_partition for non-partitioned index.
+				withClause := fmt.Sprintf(
+					" WITH { \"defer_build\": %t,  \"num_replica\": %d} ",
+					plan.With.DeferBuild.ValueBool(),
+					plan.With.NumReplica.ValueInt64(),
+				)
+
+				ddl += withClause
+			}
+
+		}
+	}
```

Comment:
In our codebase, we generally do not use many if-else statements; whenever it is required, we use a switch statement instead.